### ISSUE TITLE
Importing into Google Calendar

### ISSUE THREAD

**Issue by envycontent** [ox-it/talks.ox#366]
Event appears an hour ahead (later) when the ics is imported into Google calendar. It is interpreted fine by my Mac Calendar.

The Google calendar I'm using has the timezone set to GMT + 1 (London) as opposed to GMT (without daylight saving). 

I used Add by URL to add a series under Other Calendars (copying the URL of the add to your calendar link) 

If you think this is a Google bug, assign this to me to document. I think the default is to interpret the time without a timezone ID as local, but possibly Google takes the time as UTC (and therefore adds an hour because we are in daylight saving)

**Comment by envycontent** [comment]
Not sure if this helps, but it may be that Google can't deal with what the specs call 'floating time'

https://tools.ietf.org/html/rfc5545#section-3.3.5

**Comment by envycontent** [comment]
This might be of interest. Google seems to assume some information about the timezone - possibly this needs to be specified in the .ics (might be nice if we could supply a title as well?)

![google-calendar-settings](https://cloud.githubusercontent.com/assets/2123909/9548458/814f6d0e-4d98-11e5-8dcc-9dd7bf9d8cab.png)

**Comment by vladigor** [comment]
OK, this problem hurt my head a little bit.

TLDR version:  The system is incorrectly storing dates - it assumes they're all entered in the UTC timezone and makes no effort to localize on display, so as a self-contained site the problem isn't apparent - its only when exporting dates to an external resource that we notice the problem.

An editor might want to create a talk for, say 10am on 1st July and a talk for 10am on 1st Nov.  Both of those talks start at 10am on those days, however the July one occurs in BST so is actually at 9am UTC.  
The system stores all dates as UTC dates, but it assumes they've all been entered in the UTC timezone.
What we *want* is to set the default timezone to 'Europe/London'.  That way, when an editor saves an event for 10am on 1st July the system saves it as 9am UTC; if they save an event for 10am on 1st Nov its 10am UTC.   We then also need to tell the system to convert the time back to local timezone when displaying times on the site.

We can then tell the ics exporter to output absolute (ie. UTC) times (by appending Z to the timestamp).

This is all achievable, and I believe I've done so on my dev instance.   However, it does lead to a couple of problems.
1.  We need to make an assumption that all talks are in the Europe/London timezone.  I don't think this is a problem - all talks occur in/around Oxford, so it makes sense to enter the time in 'London' time.

2.  All talks that occur before the end of BST (ie. before 26 Oct 2015) have been incorrectly stored.  As have any that occur after the 2016 BST starts.   They need to have an hour subtracted from them.  Its probably not too difficult to run a database query to do this.
But it needs to happen at the same time as the new site is rolled out.  Otherwise the current site will display incorrect times.

**Comment by ahaith** [comment]
We should be able to add a migration step to correct existing time-zones.

**Comment by envycontent** [comment]
Fantastic! I suppose this might affect roll-out plans a bit.

re 2. It sounds like we need to make some plans as to how to do the rollout and when. If we did it after the switch over from daylight saving then that might make things slightly easier - but that may be a couple of weeks into term.  

I haven’t noticed any ‘issues’ with the talks that have been copied over to the old talks, so possibly this would affect the update to old talks as well. Given that we would probably want the new and the old talks to run alongside each other for a short while still, then we might need to look at the way in which dates are handled when copied across.


Anne Bowtell


Web Integration Manager
Medical Sciences Division

**Comment by vladigor** [comment]
I've added a migration step to fix the incorrect times and have deployed to dev.
Could you re-test importing talks into various calendars; testing talks before and after the 26 Oct.

I don't think you'll be able to run the new and old talks sites simultaneously;  once the migration step occurs on live I believe the old talks will appear to be out by an hour.

**Comment by ahaith** [comment]
We could adjust the way talks are pushed to the old site, so that, after the migration, all times are converted to the relevant timezone (ie GMT or BST as appropriate)

**Comment by envycontent** [comment]
That would be good if its feasible. I guess we will need to plan this out.

**Comment by vladigor** [comment]
That might work for any new talks being pushed to the old site, but not for any existing talks that are there   (if I understand the process correctly).

**Comment by envycontent** [comment]
Ha ha, now my head is hurting too! 

I think we're OK because the existing talks in the old system won't get touched at all unless they are updated (which only happens when a change is made in new talks) in which case a conversion on push from new talks to whatever time zone old talks (local) is using should be fine.

**Comment by vladigor** [comment]
I've just done some tests by adding talks to my dev instance and having them sync through to http://talks-ox-dev.nsms.ox.ac.uk.  
They correctly land over there with the correct time assigned, without me needing to do anything.  I'm not 100% sure how, but it seems that this isn't something we need to worry about when we have parallel sites running.

**Comment by martinfilliau** [comment]
Sounds like the iCal feed's dates are wrong because of [this function in iCal renderer](https://github.com/ox-it/talks.ox/blob/master/talks/core/renderers.py#L59-L64) : this does not specify the timezone, the `datetime` object could be passed directly [in the _event_to_ics function](https://github.com/ox-it/talks.ox/blob/master/talks/core/renderers.py#L34-L37) (as in [the official documentation](http://icalendar.readthedocs.org/en/latest/usage.html#more-documentation), assuming that the version of the `icalendar` package is correct)...

WRT tz configuration, you might want to check [USE_TZ and TIME_ZONE](https://docs.djangoproject.com/en/1.8/ref/settings/#time-zone) in the different settings file, I kind of remember that something was tricky.

My two cents, sorry :-)

**Comment by envycontent** [comment]
@vladigor This is where the old talks deals with the date - https://github.com/ox-it/talks-oxford-talks-ruby/blob/master/app/models/talk.rb#L282. 

@martinfilliau Thanks!

**Comment by envycontent** [comment]
I've checked and these now work. I'll make sure there's a ticket in the backlog about migrating dates at the switch over.

**Comment by envycontent** [comment]
Closing this ticket as the migration should have taken place.
